src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs

```diff
@@ -1102,5 +1104,218 @@ public void FormattingReturnTypeOptions()
+                         ptr.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void PublicApi_CreateOutTypeSymbol()
+        {
+            var comp = (Compilation)CreateCompilation("");
+            var @string = comp.GetSpecialType(SpecialType.System_String);
+            var @void = comp.GetSpecialType(SpecialType.System_Void);
+
+            var ptr = comp.CreateFunctionPointerTypeSymbol(
+                @void,
+                RefKind.None,
+                ImmutableArray.Create((ITypeSymbol)@string),
+                ImmutableArray.Create(RefKind.Out));
+
+            Assert.Equal("delegate*<out modreq(System.Runtime.InteropServices.OutAttribute) System.String, System.Void>",
+                         ptr.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void PublicApi_InOutAttributeMissing()
+        {
+            var comp = (Compilation)CreateCompilation("");
+            comp.MakeTypeMissing(WellKnownType.System_Runtime_InteropServices_InAttribute);
+            comp.MakeTypeMissing(WellKnownType.System_Runtime_InteropServices_OutAttribute);
+            var @string = comp.GetSpecialType(SpecialType.System_String);
+
+            var ptr = comp.CreateFunctionPointerTypeSymbol(
+                @string,
+                RefKind.RefReadOnly,
+                ImmutableArray.Create((ITypeSymbol)@string),
+                ImmutableArray.Create(RefKind.Out));
+
+            Assert.Equal("System.Runtime.InteropServices.InAttribute[missing]", ptr.Signature.RefCustomModifiers.Single().Modifier.ToTestDisplayString());
+            Assert.Equal("System.Runtime.InteropServices.OutAttribute[missing]", ptr.Signature.Parameters.Single().RefCustomModifiers.Single().Modifier.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void PublicApi_GetTypeInfo()
+        {
+            var comp = CreateFunctionPointerCompilation(@"
+unsafe class C
+{
+    public static string M1(C c) => null;
+    public static void M2(string s, int it) {}
+    public delegate*<string, int, void> M(delegate*<C, string> ptr)
+    {
+        delegate*<string, int, void> ptr2 = &M2;
+        ptr = &M1;
+        ptr(null);
+        return &M2;
+    }
+}");
+
+            comp.VerifyDiagnostics();
+
+            var syntaxTree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(syntaxTree);
+
+            var mDeclSyntax = syntaxTree.GetRoot().DescendantNodes().OfType<MethodDeclarationSyntax>().Skip(2).Single();
+
+            verifyVoidReturn(mDeclSyntax.ReturnType, useConvertedType: false);
+            verifyStringReturn(mDeclSyntax.ParameterList.Parameters[0].Type!, useConvertedType: false);
+
+            var varDecl = mDeclSyntax.DescendantNodes().OfType<VariableDeclarationSyntax>().Single();
+            verifyVoidReturn(varDecl.Type, useConvertedType: false);
```

Comment:
>verifyVoidReturn [](start = 12, length = 16)

General feedback. Extracted helpers certainly make code smaller, but they are doing this at the expense of clarity. Especially when different kinds of syntax nodes are fed to them. For example, a helper focused on a specific syntax node type would be much clearer. I am, as a reviewer, having a very hard time trying to figure out what do we test here and whether there are interesting missing scenarios.